### PR TITLE
Pandas input fn accepts DataFrame for Y

### PR DESCRIPTION
**Pull request by nrstott**
This pull request is to implement the feature request from #19182

### CODE REVIEW

**Comment by nrstott** [comment]
I signed the CLA now.

File: tensorflow/python/estimator/inputs/pandas_io.py [hunk 1]
```diff
@@ -50,7 +66,7 @@ def pandas_input_fn(x,
 
   Args:
     x: pandas `DataFrame` object.
-    y: pandas `Series` object. `None` if absent.
+    y: pandas `Series` object or `DataFrame`. `None` if absent.
```
Comment (xiejw) [inline tensorflow/python/estimator/inputs/pandas_io.py]: you probably need to update the target_column docstring also as it is not used when y is 'DataFrame'.

File: tensorflow/python/estimator/inputs/pandas_io.py [hunk 2]
```diff
@@ -117,7 +138,12 @@ def input_fn():
     features = features[1:]
     features = dict(zip(list(x.columns), features))
     if y is not None:
-      target = features.pop(target_column)
+      if isinstance(target_column, list):
```
Comment (xiejw) [inline tensorflow/python/estimator/inputs/pandas_io.py]: in order to make this check valid, it is critical to add a precheck above to ensure the target_column passed in must not be a list.

File: tensorflow/python/estimator/inputs/pandas_io.py [hunk 3]
```diff
@@ -87,7 +103,12 @@ def pandas_input_fn(x,
     if not np.array_equal(x.index, y.index):
       raise ValueError('Index for x and y are mismatched.\nIndex for x: %s\n'
                        'Index for y: %s\n' % (x.index, y.index))
-    x[target_column] = y
+    if isinstance(y, pd.DataFrame):
+      y_columns = [(column, _get_unique_target_key(x, column)) for column in list(y)]
```
Comment (xiejw) [inline tensorflow/python/estimator/inputs/pandas_io.py]: The _get_unique_target_key trick can work if y is single object not a list. Here, it is still possible to have conflicted keys. Say the keys for x is ['a'], the keys for y is ['a', 'a_n'].

**Review by xiejw — COMMENTED** [review]
Sorry for the delay. Overall the idea is great. Left some comments. Thanks!

File: tensorflow/python/estimator/inputs/pandas_io.py (hunk 1 above)
Comment (nrstott) [reply tensorflow/python/estimator/inputs/pandas_io.py]: Good point

File: tensorflow/python/estimator/inputs/pandas_io.py (hunk 2 above)
Comment (nrstott) [reply tensorflow/python/estimator/inputs/pandas_io.py]: Should it be valid for the `target_column` passed to be a list? I don't think it was valid before so this wouldn't be a behavior change so adding the precheck makes sense. Will do unless there's some compelling reason it should be able to be a list.

File: tensorflow/python/estimator/inputs/pandas_io.py (hunk 3 above)
Comment (nrstott) [reply tensorflow/python/estimator/inputs/pandas_io.py]: I see. So I need to have `_get_unique_target_key` check that the unique key isn't contained in `x` or in `y`.

**Comment by nrstott** [comment]
Thanks for the review @xiejw. Good feedback. Will get a pull request together soon to address the issues.

**Review by xiejw — COMMENTED** [review]
Sure!

**Comment by nrstott** [comment]
@xiejw Have updated the pull request to address the issues we discussed. Take a look at your convenience and let me know how it looks. Thanks!

File: tensorflow/python/estimator/inputs/pandas_io.py [hunk 4]
```diff
@@ -78,6 +96,9 @@ def pandas_input_fn(x,
     raise TypeError('shuffle must be explicitly set as boolean; '
                     'got {}'.format(shuffle))
 
+  if isinstance(target_column, list):
+    raise TypeError('target_column must be a string or None')
```
Comment (xiejw) [inline tensorflow/python/estimator/inputs/pandas_io.py]: I don't think None can work. So, this error message is not accurate. Can we just check whether target_column is six.string_types. 

For example, if user passes a dict, the error message, triggered by downstream, will be very confusing. It is better to just check against we need here. Thanks.

**Review by xiejw — COMMENTED** [review]
Thanks. Just one nit left!

File: tensorflow/python/estimator/inputs/pandas_io.py (hunk 4 above)
Comment (nrstott) [reply tensorflow/python/estimator/inputs/pandas_io.py]: Hi @xiejw, took that suggestion and am now throwing error if target_column is not an instance of six.string_types

**Review by xiejw — DISMISSED** [review]
Really appreciate the contribution here and multi-rounds of improvement.

**Comment by nrstott** [comment]
@xiejw is there a way to get more info about the failing tests? I'm seeing that they exit with error code 1. What direction should I go to debug why it's failing on the Ubuntu Python 2 and Python 3 but working on my Ubuntu machine.

**Comment by xiejw** [comment]
Based on the test log [1], if I understand the failure correctly, as six is used, you need to import six at the head of the file. You might or might not need to adjust the bazel BUILD file also. Can you try that? I can trigger the test for you at any time. 

[1] https://source.cloud.google.com/results/invocations/4b6f38a2-e310-42c8-a250-48f7c3de8568/targets/%2F%2Ftensorflow%2Fpython%2Festimator:pandas_io_test/log

**Comment by nrstott** [comment]
@xiejw Thanks, that makes complete sense. I'm now confused though as to why the test does not produce an error on my local. It looks like it should! I did a bazel clean before trying it again and it still passes even though six is not imported. This is the command I'm using:

```bazel test --config=opt //tensorflow/python/estimator:pandas_io_test```

I've also used

```bazel run --config=opt //tensorflow/python/estimator:pandas_io_test ```

It says 17 passing.

Should I be running the tests in a different way for local development?

**Comment by xiejw** [comment]
I think given the tests need to be run against different python versions and platforms. The best way to try it is to submit your commit and I will kick another test run for you.

**Comment by nrstott** [comment]
Ok submitted the commit. Kick off the tests when ready.

**Review by xiejw — DISMISSED** [review]

**Review by xiejw — DISMISSED** [review]

**Comment by xiejw** [comment]
I think the final error is easy to fix. 

tensorflow/python/estimator/inputs/pandas_io.py:114: [C0301(line-too-long), ] Line too long (85/80)

**Review by xiejw — DISMISSED** [review]

**Comment by xiejw** [comment]
@yifeif  Yifei, is there a link or page describing how to run lint in open source. It is annoying for contributors to keep committing and trying here (just for a trivial lint error)

**Comment by yifeif** [comment]
For future references, you can run https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/ci_sanity.sh locally for the sanity check.

**Comment by nrstott** [comment]
Thanks @yifeif. I ran that locally and got some failures that don't look like they are related to my changes. Is my environment not setup correctly? 

```bash
4. do_buildifier: buildifier check
  FAIL
5. do_bazel_nobuild: bazel nobuild
  FAIL
6. do_pip_package_licenses_check: pip: license check for external dependencies
  FAIL
7. do_lib_package_licenses_check: C library: license check for external dependencies
  FAIL
8. do_java_package_licenses_check: Java Native Library: license check for external dependencies
  FAIL
9. do_pip_smoke_test: Pip Smoke Test: Checking py_test dependencies exist in pip package
  FAIL
10. do_check_load_py_test: Check load py_test: Check that BUILD files with py_test target properly load py_test
  FAIL
```

**Review by xiejw — APPROVED** [review]

**Pull request merged**